eta/core/datasets.py

```diff
@@ -1338,3 +1375,842 @@ def required(cls):
+
+        Returns: Integer target value
+        '''
+        if self.target_count:
+            return self.target_count
+        return int(np.quantile(counts, self.target_quantile))
+
+    def _get_keep_idxs(self, A, counts, target_count):
+        '''Algorithm fun! This function chooses the set of records to keep (and
+        remove).
+
+        There's still plenty of potential for testing and improvement here.
+
+
+        This problem can be posed as:
+            minimize:
+                ||np.dot(A, x) - b||
+            subject to:
+                x[i] is an element of [0, 1]
+
+        and different algorithms may be substituted in.
+
+        Args:
+            A: the occurrence matrix computed in Balancer._get_occurrence_matrix
+            counts (vector): the original counts for each value
+            target_count (int): the target value
+
+        Returns:
+            a list of integer indices to keep
+        '''
+        b = counts - target_count
+
+        x = self._greedy(A, b)
+
+        return np.where(x == 0)[0]
+
+    def _random(self, A, b):
+        '''A random search algorithm for finding the indices to omit.
+
+        Args:
+            A: the occurrence matrix
+            b: the target vector to match
+
+        Returns:
+            x: the solution vector. x[j]=1 -> omit the j'th record
+        '''
+        best_x = np.zeros(A.shape[1], dtype=np.dtype("int"))
+        best_score = self._solution_score(b - np.dot(A, best_x))
+
+        random.seed(1)
+        for _ in range(10000):
+            i = random.choice(np.where(best_x == 0)[0])
+            cur_x = best_x.copy()
+            cur_x[i] = 1
+            cur_score = self._solution_score(b - np.dot(A, cur_x))
+            if cur_score < best_score:
+                best_score = cur_score
+                best_x = cur_x
+
+        return best_x
+
+    def _greedy(self, A, b):
+        '''A greedy search algorithm for finding the indices to omit.
+
+        Args:
+            A: the occurrence matrix
+            b: the target vector to match
+
+        Returns:
+            x: the solution vector. x[j]=1 -> omit the j'th record
+        '''
+        best_x = np.zeros(A.shape[1], dtype=np.dtype('int'))
+        best_score = self._solution_score(b - np.dot(A, best_x))
+        w = np.where(best_x == 0)[0]
+
+        while len(w):
+            x_matrix = np.zeros((len(best_x), len(w)), dtype=np.dtype('int'))
+            for idx, val in enumerate(w):
+                x_matrix[:, idx] = best_x
+                x_matrix[val, idx] = 1
+            Y = np.expand_dims(b, axis=1) - np.dot(A, x_matrix)
+
+            cur_scores = self._solution_score(Y.T)
+
+            i = np.argmin(cur_scores)
+
+            if cur_scores[i] >= best_score:
+                break
+
+            best_score = cur_scores[i]
+            best_x = x_matrix[:, i]
+            w = np.where(best_x == 0)[0]
+
+        return best_x
+
+    def _solution_score(self, vector):
```

Comment:
same comment about supporting customization of this function, and refactoring this implementation into a top-level function to avoid polluting the definition of this class